presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -222,18 +214,35 @@ else if (canQueueMore()) {
     {
         synchronized (root) {
             if (subGroups.isEmpty()) {
+                List<ResourceGroupInfo> pathToRoot = getPathToRoot();
```

Comment:
This implementation definitely looks cleaner than the previous one. The only difference is that this logic will not be in a synchronized block in the previous implementation, so there is potentially a performance difference. I'm leaning towards that this performance difference does not matter and it's worth it for the simplicity. But I want to point it out so you are aware of this choice.

Reply:
Again, thanks for the reminder. I also believe the performance will be OK.